WT-9705 Fix accidentally freeing update on the update chain in prepared commit

quchenhao opened the pull request:
Before this change, we were using the fix_upd pointer to pass the timestamp information to the hs fixup logic. I reorganized the code so that we don't need to rely on that pointer. In this way, we don't need to worry whether we need to free the fix_upd which leads to this issue.

quchenhao commented:
@agorrod Can you try to reproduce your issue with this fix?

quchenhao commented:
@kommiharibabu @agorrod  This is ready to review.

kommiharibabu commented on test/suite/test_prepare24.py:
```diff
@@ -0,0 +1,125 @@
+
+# test_prepare24.py
+# Test prepare commit after eviction failure.
+class test_prepare23(wttest.WiredTigerTestCase):
```
The class name should match the test name. `prepare24`.

quchenhao replied:
done

kommiharibabu commented on test/suite/test_prepare24.py:
```diff
@@ -0,0 +1,125 @@
+            # Insert a value
+            self.session.begin_transaction()
+            cursor[i] = value_b
+            self.session.prepare_transaction('prepare_timestamp=' + self.timestamp_str(ts + 30))
```
The cursor is not closed, so it can pin the page where the eviction wants to evict it.

quchenhao replied:
done

kommiharibabu reviewed (approved):
LGTM

agorrod commented:
@quchenhao @bvpvamsikrishna and I had a conversation about this change. We discussed whether there was something we could do to more clearly define the relationship between the two parts of the code at play in this bug.

Siddhartha8899 commented on src/txn/txn.c:
```diff
@@ -926,27 +901,33 @@ __txn_timestamp_usage_check(WT_SESSION_IMPL *session, WT_TXN_OP *op, WT_UPDATE *
-    uint32_t txn_flags;
-#ifdef HAVE_DIAGNOSTIC
-    uint64_t hs_upd_type;
     wt_timestamp_t hs_durable_ts, hs_stop_durable_ts;
```
We can get rid of these?

quchenhao replied:
Why? They are used later.

Siddhartha8899 replied:
Thanks for clarifying that we cannot pass NULL in the `get_value()` function. Resolving this comment.

quchenhao commented on src/txn/txn.c:
```diff
@@ -1167,18 +1149,17 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
     txn = session->txn;
-    fix_upd = NULL;
-    upd_appended = false;
+    prepare_on_disk = false;
```
yes

```
prepare_verify:
#ifdef HAVE_DIAGNOSTIC
    for (; head_upd != NULL; head_upd = head_upd->next) {
        /*
         * Assert if we still have an update from the current transaction that hasn't been resolved
         * or aborted.
         */
        WT_ASSERT(session,
          head_upd->txnid == WT_TXN_ABORTED || head_upd->prepare_state == WT_PREPARE_RESOLVED ||
            head_upd->txnid != txn->id);

        if (head_upd->txnid == WT_TXN_ABORTED)
            continue;

        /*
         * If we restored an update from the history store, it should be the last update on the
         * chain.
         */
        if (!commit && prepare_on_disk && head_upd->type == WT_UPDATE_STANDARD &&
          F_ISSET(head_upd, WT_UPDATE_RESTORED_FROM_HS))
            WT_ASSERT(session, head_upd->next == NULL);
    }
#endif
```
Otherwise it goes there unintialized

Siddhartha8899 replied:
Thanks, makes sense.

Siddhartha8899 reviewed (approved):
LGTM!

quchenhao commented:
@agorrod I haven't had a good idea to restructure the code. Therefore, I added more asserts and rewrote some comments. Hopefully that makes the logic clearer.

quchenhao commented:
Before merging, I'll disable the fail point again until all the other fixes are in.

quchenhao commented on src/txn/txn.c:
```diff
@@ -717,14 +717,12 @@ __wt_txn_release(WT_SESSION_IMPL *session)
-__txn_locate_hs_record(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_PAGE *page,
-  WT_UPDATE *chain, bool commit, WT_UPDATE **fix_updp, bool *upd_appended,
-  WT_UPDATE *first_committed_upd, bool first_committed_upd_in_hs)
+__txn_restore_hs_update(
```
This function now only does one thing to restore the history store update to the update chain if we have a prepared rollback.

quchenhao replied:
The update we allocate here in this function should be always append to the update chain. Therefore, there is no need to free the update anymore. As in the previous implementation, we may allocate a temporary update but not append it to pass on the timestamp information to the later code.

quchenhao commented on src/txn/txn.c:
```diff
@@ -964,32 +953,30 @@ __txn_fixup_prepared_update(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_U
     tw.durable_stop_ts = txn->durable_timestamp;
     tw.stop_txn = txn->id;
-    WT_TIME_WINDOW_SET_START(&tw, fix_upd);
+    tw.start_ts = hs_tw->start_ts;
```
Use timestamp from the history store cell instead of using an update to pass the information.

quchenhao commented on src/txn/txn.c:
```diff
@@ -1231,20 +1217,39 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
-     * WT_UPDATE_HS. The second case is rare but can happen if the eviction that writes the prepared
-     * update to the disk image fails after it has inserted the other updates of the key into the
-     * history store.
+     * If the prepared update is a single tombstone, we don't need to do anything special and we can
```
Add comments to describe all the scenario.

quchenhao commented on src/txn/txn.c:
```diff
@@ -1387,8 +1420,6 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
 err:
     if (hs_cursor != NULL)
         WT_TRET(hs_cursor->close(hs_cursor));
-    if (!upd_appended)
```
No need to free anything.

agorrod commented on src/txn/txn.c:
```diff
@@ -964,32 +953,30 @@ __txn_fixup_prepared_update(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_U
     tw.durable_stop_ts = txn->durable_timestamp;
     tw.stop_txn = txn->id;
-    WT_TIME_WINDOW_SET_START(&tw, fix_upd);
+    tw.start_ts = hs_tw->start_ts;
```
It feels like maybe we need to expand the set of time window initializers? Ideally there would be a single call with the start coming from hs_tw and stop coming from txn.
I don't think it needs to be a macro - could be an inline function.

quchenhao replied:
Added a macro to do that.

agorrod reviewed (commented):
I've left a few comments, hopefully they make sense.

quchenhao commented on src/txn/txn.c:
```diff
@@ -964,32 +953,30 @@ __txn_fixup_prepared_update(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_U
+    tw.durable_start_ts = hs_tw->durable_start_ts;
+    tw.start_txn = hs_tw->start_txn;
 
-#ifdef HAVE_DIAGNOSTIC
```
We can still can that check. I removed it because we already checked this before

```
    /*
     * If the history update already has a stop time point there is no work to do. This happens if a
     * deleted key is reinserted by a prepared update.
     */
    if (WT_TIME_WINDOW_HAS_STOP(hs_tw))
        return (0);
```

quchenhao replied:
Fixed

quchenhao commented on src/txn/txn.c:
```diff
@@ -1291,6 +1313,18 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
+         * sessions modifying the same key and checkpoint moving the new updates to the history
+         * store.
+         */
         WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_before(session, hs_cursor), true);
```
I restructured the code. Hopefully it's now clearer

agorrod replied:
I meant to say - much better now - thanks!

quchenhao commented on src/txn/txn.c:
```diff
@@ -1253,6 +1258,23 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
     first_committed_upd_in_hs =
       first_committed_upd != NULL && F_ISSET(first_committed_upd, WT_UPDATE_HS);
 
+    /* We should not see the flags both set. */
```
Fixed

agorrod commented on src/include/timestamp_inline.h:
```diff
@@ -65,6 +65,22 @@
     } while (0)
 
+/* Copy the start values of a time window from another time window. */
+#define WT_TIME_WINDOW_COPY_START(tw1, tw2)                \
```
I prefer `source` and `dest` rather than `tw1` and `tw2` to be clear which direction the initialization happens in. I also wonder if there is a reason to use a macro rather than an inline function? I guess we'd need two (or four) inline functions, to deal with different data structures with time windows - though I feel like that type checking makes the duplication worthwhile.

quchenhao replied:
The update and time window they have different fields in the data structure so we would still get a compile failure when we pass the wrong type. I use macro just to be consistent to what we already have.

agorrod reviewed (commented):
Thanks @quchenhao - this looks better to me now. I've got one more question/request.

Could you please ensure a round of stress testing has completed before this gets merged?

quchenhao commented:
stress testing:
https://spruce.mongodb.com/version/6305906cd1fe0772d286302f/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

mongodb patch build:
https://spruce.mongodb.com/version/630591b2d1fe0772d286312e/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

agorrod reviewed (approved):
Thanks @quchenhao  - I'm happy with this now if it passes stress testing.

bvpvamsikrishna commented on src/txn/txn.c:
```diff
@@ -1253,6 +1258,26 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
+
     /*
      * Marked the update older than the prepared update that is already in the history store to be
      * deleted from the history store.
```
I find it difficult to follow the below logic.
At this stage if first_committed_upd_in_hs is true, then first_committed_upd is from history store, why we need to loop to locate the upd_followed_tombstone and check whether its txn_id != WT_TXN_ABORTED, as these updates are in history store or restored from history store, in both cases there will be no aborted update.

The logic below line 1301, trying to find the latest history store entry and trying to restore it, but that might not be required if we already have the first_committed_upd.

May be we can structure this code as switch: case to make it clear which cases is being handled based on the cases described in the above comment at line 1220?

quchenhao replied:
At this stage if first_committed_upd_in_hs is true, then first_committed_upd is from history store, why we need to loop to locate the upd_followed_tombstone and check whether its txn_id != WT_TXN_ABORTED, as these updates are in history store or restored from history store, in both cases there will be no aborted update.

No, in this case we are only handling the case that things are in history store but not restored from history store. We can have aborted updates in the middle as they are never successfully evicted.

The logic below line 1301, trying to find the latest history store entry and trying to restore it, but that might not be required if we already have the first_committed_upd.

We already check that by `!commit && prepare_on_disk`, we are not restoring from hs in case of `first_committed_upd_in_hs`

May be we can structure this code as switch: case to make it clear which cases is being handled based on the cases described in the above comment at line 1220?

It's a bit hard to do that. We will create some code duplication as a lot of the steps are shared.

quchenhao replied:
After a second thought, I think this is a good idea. Done as suggested.

quchenhao commented on src/txn/txn.c:
```diff
@@ -964,32 +955,28 @@ __txn_fixup_prepared_update(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_U
     WT_ERR(hs_cursor->update(hs_cursor));
 
 err:
     F_SET(txn, txn_flags);
```
We should clear if it is not set originally

quchenhao replied:
fixed

quchenhao commented on src/txn/txn.c:
```diff
@@ -717,14 +717,12 @@ __wt_txn_release(WT_SESSION_IMPL *session)
-__txn_locate_hs_record(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_PAGE *page,
-  WT_UPDATE *chain, bool commit, WT_UPDATE **fix_updp, bool *upd_appended,
-  WT_UPDATE *first_committed_upd, bool first_committed_upd_in_hs)
+__txn_restore_hs_update(
```
fixed

quchenhao commented on src/txn/txn.c:
```diff
@@ -1291,23 +1333,42 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
-        if (ret == WT_NOTFOUND && !commit) {
+        WT_ASSERT(session, ret != WT_NOTFOUND || resolve_case == RESOLVE_PREPARE_ON_DISK);
+        if (ret == 0) {
+            has_hs_record = true;
```
This fixes the issue in the previous version that we will fix the history store record even there is none history store record.

quchenhao commented:
new stress test
https://spruce.mongodb.com/version/63060854c9ec447fb468281b/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

quchenhao commented:
pr testing failure is a known issue

quchenhao commented:
evergreen merge
